fdbserver/workloads/DcLag.actor.cpp

```diff
@@ -0,0 +1,188 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbclient/NativeAPI.actor.h"
+#include "fdbclient/StatusClient.h"
+#include "fdbrpc/Locality.h"
+#include "fdbrpc/SimulatorProcessInfo.h"
+#include "fdbserver/Knobs.h"
+#include "fdbserver/RecoveryState.h"
+#include "fdbserver/ServerDBInfo.actor.h"
+#include "fdbserver/TesterInterface.actor.h"
+#include "fdbserver/workloads/workloads.actor.h"
+#include "fdbrpc/simulator.h"
+#include "flow/CodeProbe.h"
+#include "flow/NetworkAddress.h"
+#include "flow/Error.h"
+#include "flow/Trace.h"
+#include "flow/flow.h"
+#include "flow/network.h"
+#include "flow/actorcompiler.h" // This must be the last #include.
+
+// This workload tests a gray failure scenario: a satellite TLog is have network issue
+// for sending packets to the remote data center's log routers. This will cause these
+// log routers to fail to progress, and causing data center lag, i.e., lag between the
+// primary and remote DC's tlags. With changes to log routers, they can detect that the
+// peek is taking a long time (> LOG_ROUTER_PEEK_SWITCH_DC_TIME) and swith to use
+// another DC to get data, thus recoverying from the data center lag.
+struct DcLagWorkload : TestWorkload {
+	static constexpr auto NAME = "DcLag";
+	bool enabled;
+	double testDuration;
+	double startDelay;
+	std::vector<std::pair<IPAddress, IPAddress>> cloggedPairs;
+
+	DcLagWorkload(WorkloadContext const& wcx) : TestWorkload(wcx) {
+		enabled = !clientId; // only do this on the "first" client
+		testDuration = getOption(options, "testDuration"_sr, 1000.0);
+		startDelay = getOption(options, "startDelay"_sr, 10.0);
+	}
+
+	void disableFailureInjectionWorkloads(std::set<std::string>& out) const override { out.insert("Attrition"); }
+
+	Future<Void> setup(Database const& cx) override { return Void(); }
+	Future<Void> start(Database const& cx) override {
+		if (g_network->isSimulated() && enabled)
+			return timeout(reportErrors(clogClient(this, cx), "DcLagError"), testDuration, Void());
+		else
+			return Void();
+	}
+	Future<bool> check(Database const& cx) override { return true; }
+	void getMetrics(std::vector<PerfMetric>& m) override {}
+
+	// Clog a satellite tlog with all remote processes so that this triggers high
+	// data center lag.
+	bool clogTlog(double seconds) {
+		ASSERT(dbInfo->get().recoveryState >= RecoveryState::RECOVERY_TRANSACTION);
+
+		std::vector<IPAddress> ips; // all remote process IPs
+		for (const auto& process : g_simulator->getAllProcesses()) {
+			const auto& ip = process->address.ip;
+			if (process->locality.dcId().present() && process->locality.dcId().get() == g_simulator->remoteDcId) {
+				ips.push_back(ip);
+			}
+		}
+		ASSERT(ips.size() > 0);
+
+		// Find all satellite tlogs
+		std::vector<NetworkAddress> logs; // all satellite logs
+		for (int i = 1; i < dbInfo->get().logSystemConfig.tLogs.size(); i++) {
```

Comment:
Consider adding a comment to explain why i starts from 1.

Reply:
Also, maybe consider using `tlogset.locality == tagLocalitySatellite`? Seems more readable.

Reply:
Simplify the code and thanks for the suggestion. `tlogset.isLocal` below is used to find primary satellite. we don't want to include any remote satellite processes.